README

A library to compute Jacobi polynomials. Version 0.9.2 (2006-11-22)

This library provides functions that can compute Jacobi Polynomials.
It also provides functions to deal with Gauss-Jacobi quadratures. It 
can calculate the nodes and weights. It can also compute the derivative
matrix and interpolation matrix. It is based on code from the C 
library libjacobi. The only difference is that any dependency on 
the GNU Scientific library has been removed.

Paulo Jabardo <pjabardo@yahoo.com.br>



NEWS
2006-11-22 Version 0.9.2 is out. Improved the interface a little
2006-07-24 Version 0.9.1 is out. Removed every warning of R CMD check 
2006-07-18 Version 0.9 is out


R/rjacobi.R



jacobi <- function(x,n,a=0.0, b=0.0){
  .Call("R_jacobi", x, n, a, b, PACKAGE='rjacobi')
}

djacobi <- function(x,n,a=0.0, b=0.0){
  .Call("R_djacobi", x, n, a, b, PACKAGE='rjacobi')
}


jacobiZeros <- function(Q, a=0, b=0)
  .Call("R_jacobi_zeros", Q, a, b, PACKAGE='rjacobi')


zerosGJ <- function(Q, a=0, b=0)
  .Call("R_zeros_gj", Q, a, b, PACKAGE='rjacobi')

zerosGLJ <- function(Q, a=0, b=0)
  .Call("R_zeros_glj", Q, a, b, PACKAGE='rjacobi')

zerosGRJM <- function(Q, a=0, b=0)
  .Call("R_zeros_grjm", Q, a, b, PACKAGE='rjacobi')

zerosGRJP <- function(Q, a=0, b=0)
  .Call("R_zeros_grjp", Q, a, b, PACKAGE='rjacobi')


weightsGJ <- function(x, a=0,b=0)
  .Call("R_weights_gj", x, a, b, PACKAGE='rjacobi')

weightsGLJ <- function(x,a=0,b=0)
  .Call("R_weights_glj", x, a, b, PACKAGE='rjacobi')

weightsGRJM <- function(x,a=0,b=0)
  .Call("R_weights_grjm", x, a, b, PACKAGE='rjacobi')

weightsGRJP <- function(x,a=0,b=0)
  .Call("R_weights_grjp", x, a, b, PACKAGE='rjacobi')


diffmatGJ <- function(x,a=0, b=0)
  t(.Call("R_diffmat_gj", x, a, b, PACKAGE='rjacobi'))

diffmatGLJ <- function(x,a=0, b=0)
  t(.Call("R_diffmat_glj", x, a, b, PACKAGE='rjacobi'))

diffmatGRJM <- function(x,a=0, b=0)
  t(.Call("R_diffmat_grjm", x, a, b, PACKAGE='rjacobi'))

diffmatGRJP <- function(x,a=0, b=0)
  t(.Call("R_diffmat_grjp", x, a, b, PACKAGE='rjacobi'))



interpmatGJ <- function(zp, x, a=0, b=0)
  t(.Call("R_interpmat_gj", zp, x, a, b, PACKAGE='rjacobi'))

interpmatGLJ <- function(zp, x, a=0, b=0)
  t(.Call("R_interpmat_glj", zp, x, a, b, PACKAGE='rjacobi'))

interpmatGRJM <- function(zp, x, a=0, b=0)
  t(.Call("R_interpmat_grjm", zp, x, a, b, PACKAGE='rjacobi'))

interpmatGRJP <- function(zp, x, a=0, b=0)
  t(.Call("R_interpmat_grjp", zp, x, a, b, PACKAGE='rjacobi'))





# Wrappers of the functions above to ease on the use of the library
quadGJ <- function(Q, a=0, b=0){
  x <- zerosGJ(Q, a, b)
  w <- weightsGJ(x, a, b)
  D <- diffmatGJ(x, a, b)

  # Interpolation matrix, build an abstraction
  interp <- function(zp)
    interpmatGJ(zp, x, a, b)

  return(list(x=x, w=w, D=D, interp=interp))
}

quadGLJ <- function(Q, a=0, b=0){
  x <- zerosGLJ(Q, a, b)
  w <- weightsGLJ(x, a, b)
  D <- diffmatGLJ(x, a, b)

  # Interpolation matrix, build an abstraction
  interp <- function(zp)
    interpmatGLJ(zp, x, a, b)

  return(list(x=x, w=w, D=D, interp=interp))
}


quadGRJM <- function(Q, a=0, b=0){
  x <- zerosGRJM(Q, a, b)
  w <- weightsGRJM(x, a, b)
  D <- diffmatGRJM(x, a, b)

  # Interpolation matrix, build an abstraction
  interp <- function(zp)
    interpmatGRJM(zp, x, a, b)

  return(list(x=x, w=w, D=D, interp=interp))
}

         
quadGRJP <- function(Q, a=0, b=0){
  x <- zerosGRJP(Q, a, b)
  w <- weightsGRJP(x, a, b)
  D <- diffmatGRJP(x, a, b)

  # Interpolation matrix, build an abstraction
  interp <- function(zp)
    interpmatGRJP(zp, x, a, b)

  return(list(x=x, w=w, D=D, interp=interp))
}

quadrature <- function(Q, type='GJ', a=0, b=0, xinterp=NULL){
  type <- match.arg(toupper(type), c('GJ', 'GLJ', 'GRJM', 'GRJP'))

  quadfun <- get(paste('quad', type, sep=''))
  q <- quadfun(Q, a, b)
  q$a <- a
  q$b <- b
  q$type <- type
  
  lagr <- get(paste('lagrange', type, sep=''))
  lagrange <- function(x,i)
    lagr(x, i, q$x, q$a, q$b)

  q$lagrange <- lagrange
  if (!is.null(xinterp)){
    q$xinterp <- xinterp
    q$imat <- q$interp(xinterp)
  }else{
    q$xinterp <- NULL
    q$imat <- NULL
  }

  class(q) <- "quadrature"
  return(q)
}


xinterp <- function(q){
  return(q$xinterp)
}

'xinterp<-' <- function(q, value){
  q$xinterp <- value
  q$imat <- q$interp(value)
  return(q)
}

quadIntegr <- function(f,quad){
  if (is.function(f)) f <- f(quad$x)

  return(sum(f*quad$w))
}

quadDeriv <- function(f, quad){
  if (is.function(f)) f <- f(quad$x)

  return(quad$D %*% f)
}

quadInterp <- function(imat, f){
  if (class(imat) == 'quadrature'){
    if (is.function(f)) f <- f(imat$x)
    imat <- imat$imat
  }
  
  return(imat %*% f)
}


lagrangeGJ <- function(z, i, x, a=0.0, b=0.0){
  .Call("R_lagrange_gj", i, z, x, a, b, PACKAGE='rjacobi')
}


lagrangeGLJ <- function(z, i, x, a=0.0, b=0.0){
  .Call("R_lagrange_glj", i, z, x, a, b, PACKAGE='rjacobi')
}


lagrangeGRJM <- function(z, i, x, a=0.0, b=0.0){
  .Call("R_lagrange_grjm", i, z, x, a, b, PACKAGE='rjacobi')
}


lagrangeGRJP <- function(z, i, x, a=0.0, b=0.0){
  .Call("R_lagrange_grjp", i, z, x, a, b, PACKAGE='rjacobi')
}



src/jacobi.c
/* jacobi.c
 * Copyright (C) 2006 Paulo José Saiz Jabardo
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

/* Author:  Paulo J. Saiz Jabardo */






/** \file jacobi.c
 
 \brief Calculation of Jacobi Polynomials

 This file implements the function that calculates Jacobi polynomials and its derivatives.
 It tries to follow the interfaces used in the GNU Scientific library. The estimated errors
 Have not been implemented yet.

  The main sources used to develop this library where:

  (1) Abramowitz, Milton and Stegun, Irene (editors); "Handbook of Mathematical functions",
      Dover Publications 1965.

  (2) Karniadakis, George Em and Sherwin, Spencer; "Spectral/hp Element Methods for Computational
      Fluid Dynamics", Oxford Science Publications, 2nd edition, 2005.


   The Jacobi polynomials of order n are defined as follows

   \f[ P_n^{\alpha,\beta}(x) \f]

   The calculation uses the recursive definition of the Jacobi polynomial:

   \f[
   a^1_n P_{n+1}^{\alpha,\beta}(x) = (a^2_n + a^3_n x)P_n^{\alpha,\beta}(x) - a^4_n P_{n-1}^{\alpha,\beta}(x)
   \f]
   
   In this function, \f$-1\le x \le 1\f$ is the point where the function is to be calculated, n is the order of the
   polynomial and \f$a^i_n\f$ are coefficients given by

   \f[
     a^1_n = 2(n+1)(n+\alpha+\beta+1)(2n+\alpha+\beta)
   \f]
   \f[
     a^2_n = (2n+\alpha+\beta+1)(\alpha^2-\beta^2)
   \f]
   \f[
     a^3_n = (2n + \alpha + \beta)(2n + \alpha + \beta + 1)(2n + \alpha +\beta + 2)
   \f]
   \f[
     a^4_n = 2(n+\alpha)(n+\beta)(2n+\alpha+\beta+2)
   \f]



   There is also a recursive definition of the derivative of Jacobi polynomials but in this library, the derivatives
   of Jacobi polynomials will be defined in terms of Jacobi polynomials:

   \f[
     \frac{d}{dx}P^{\alpha,\beta}_n(x) = \frac{1}{2}(\alpha + \beta + n + 1) P_n^{\alpha+1,\beta+1}(x)
   \f]

   Chebychev polynomials are a special case of Jacobi polynomials with \f$\alpha,\beta=-1/2\f$:

   \f[
      T_n(x) = \frac{2^{2n}(n!)^2}{(2n)!} P_n^{-1/2,-1/2}(x)
   \f]

   Since the Zeros of Chebychev are explicitly known, an iteration using these zeros as initial guess is used
   to calculate zeros of generic Jacobi polynomials.
   

 */    
 
 


#include <stdlib.h>


#include "gsl_utils.h"



#include "jacobi.h"





  

/** Calculates the 0th order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param  result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
  
 */
int
jac_jacobi_P0_e (double x, double a, double b, gsl_sf_result *result)
{
    result->val = 1.0;
    result->err = 0.0;
    return GSL_SUCCESS;
}

/** Calculates the 1st order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
  
 */
int
jac_jacobi_P1_e (double x, double a, double b, gsl_sf_result *result)
{
    result->val = 0.5 * (a - b + (a + b + 2.0) * x);
    result->err = 0.0;
    return GSL_SUCCESS;
}



/** Calculates the nth order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param n An integer specifying the order of the polynomial
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
 
 */
int
jac_jacobi_e (double x, int n, double a, double b, gsl_sf_result *result)
{
    
    if (n==0)
    {
	result->val = 1.0;
	result->err = 0.0;
	return GSL_SUCCESS;
    }
    else if (n==1)
    {
	result->val = 0.5 * (a - b + (a + b + 2.0)*x);
	result->err = 0.0;
	return GSL_SUCCESS;
    }
    else
    {
	
	double p0, p1, a1, a2, a3, a4, p2;
	int i;
	p0 = 1.0;
	p1 = 0.5 * (a - b + (a + b + 2)*x);
    
	for(i=1; i<n; ++i){
	    a1 = 2.0*(i+1.0)*(i+a+b+1.0)*(2.0*i+a+b);
	    a2 = (2.0*i+a+b+1.0)*(a*a-b*b);
	    a3 = (2.0*i+a+b)*(2.0*i+a+b+1.0)*(2.0*i+a+b+2.0);
	    a4 = 2.0*(i+a)*(i+b)*(2.0*i+a+b+2.0);
	    p2 = 1.0/a1*( (a2 + a3*x)*p1 - a4*p0);
	
	    p0 = p1;
	    p1 = p2;
	}
    
	result->val = p2;
	result->err = 0.0;
	return GSL_SUCCESS;
    }
}


/** Explicit calculation of the 0th order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_jacobi_P0(double x, double a, double b)
{
    return 1.0;
}

/** Explicit calculation of the 1st order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_jacobi_P1(double x, double a, double b)
{
    return 0.5 * (a - b + (a + b + 2.0) * x);
}




/** Explicit calculation of the nth order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param n An integer specifying the order of the polynomial
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_jacobi (double x, int n, double a, double b)
{
    
    if (n==0)
    {
	return 1.0;
    }
    else if (n==1)
    {
	return  0.5 * (a - b + (a + b + 2.0)*x);
    }
    else
    {
	
	double p0, p1, a1, a2, a3, a4, p2;
	int i;
	p0 = 1.0;
	p1 = 0.5 * (a - b + (a + b + 2)*x);
    
	for(i=1; i<n; ++i){
	    a1 = 2.0*(i+1.0)*(i+a+b+1.0)*(2.0*i+a+b);
	    a2 = (2.0*i+a+b+1.0)*(a*a-b*b);
	    a3 = (2.0*i+a+b)*(2.0*i+a+b+1.0)*(2.0*i+a+b+2.0);
	    a4 = 2.0*(i+a)*(i+b)*(2.0*i+a+b+2.0);
	    p2 = 1.0/a1*( (a2 + a3*x)*p1 - a4*p0);
	
	    p0 = p1;
	    p1 = p2;
	}
    
	return p2;
    }
}





/** Calculates the derivative of the 0th order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
  
 */
int
jac_djacobi_P0_e (double x, double a, double b, gsl_sf_result *result)
{
    result->val = 0.0;
    result->err = 0.0;
    return GSL_SUCCESS;
}

/** Calculates the derivative of the 1st order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
  
 */
int
jac_djacobi_P1_e (double x, double a, double b, gsl_sf_result *result)
{
    result->val = 0.5 * (a + b + 2.0);
    result->err = 0.0;
    return GSL_SUCCESS;
}


/** Calculates the derivative of the nth order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param n Integer specifying the order of the polynomial
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \param result A pointer to a gsl_sf_result structure where the result and estimated error is stored
  \return GSL_SUCCESS if successful
  
 */
int
jac_djacobi_e (double x, int n, double a, double b, gsl_sf_result *result)
{
    int code = jac_jacobi_e(x, n-1, a+1.0, b+1.0, result);
    
    result->val *= 0.5  * (a + b + n + 1.0);
    result->err = 0.0;
    return code;

}


/** Explicit calculation of the derivative of the 0th order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_djacobi_P0(double x, double a, double b)
{
    return 0.0;
}

/** Explicit calculation of the derivative of the 1st order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_djacobi_P1(double x, double a, double b)
{
    return 0.5 * (a + b + 2.0);
}


/** Explicit calculation of the derivative of the nth order Jacobi polynomial

  \param x The point where the function should be calculated. \f$-1\le x \le 1\f$
  \param n The order of the polynomial
  \param a \f$\alpha\f$ parameter of the Jacobi Polynomial
  \param b \f$\beta\f$ parameter of the Jacobi Polynomial
  \return The calculated value of the Jacobi polynomial
  
 */
double
jac_djacobi (double x, int n, double a, double b)
{
    
    if (n==0)
    {
	return 0.0;
    }
    else if (n==1)
    {
	return  0.5 * (a + b + 2.0);
    }
    else
    {
	return 0.5 * (a + b + n + 1.0) * jac_jacobi(x, n-1, a+1.0, b+1.0);
    }
}



/** Calculates the nth order Jacobi polynomials at an array of np points given by the vector x.
    This function needs two workspaces of size np to do its calculation. If the first one (ws1) is
    NULL, the function will allocate the necessary memory (using malloc) to do the calculation

    \param np Number of points where the polynomial will be calculated
    \param x A vector of length np containing the points where the polynomials should be calculated. \f$-1\le x_i \le 1\f$.
    \param n Order of Jacobi polynomial to be calculated
    \param result_array An array of length np where the result will be stored
    \param a \f$\alpha\f$ parameter of Jacobi polynomial
    \param b \f$\beta\f$ parameter of Jacobi polynomial
    \param ws A pointer to a block of memory at least 2*np doubles long to be used as workspace. If it is NULL, memory will be allocated using malloc and released at the end
    \return Error code defined in gsl_errno.h or GSL_SUCCESS if everything was fine
 */
int 
jac_jacobi_array (int np, const double *x, int n, double * result_array,
		  double a, double b, double *ws)
{



    double *pnm1, *pnm2;
    int mem_allocated=0;
    int i;

    int k;
    double a1, a2, a3, a4;

    /* Computes the jacobi polyniomials on several points*/
    if (n == 0)
    {
	for(i = 0; i < np; ++i)
	    result_array[i] = 1.0;
	return GSL_SUCCESS;
    }
    
    if (n == 1)
    {
	for (i = 0; i < np; ++i)
	    result_array[i] = 0.5 * (a - b + (a + b + 2.0)*x[i]);
	return GSL_SUCCESS;
    }
    
    /* General case,*/
    if (ws==NULL)
    {
	pnm1 = (double *) malloc(2*np*sizeof(double));
	if (!pnm1) return GSL_ENOMEM;
	
	mem_allocated = 1;
	pnm2 = pnm1 + np;
    } else
    {
	pnm1 = ws;
	pnm2 = ws + np;
    }
    
    for(i = 0; i < np; ++i){
	pnm2[i] = 1.0;
	pnm1[i] = 0.5 * (a - b + (a + b + 2.0)*x[i]);
    }
    
    /* Start iterating:*/
    for(k=1; k<n; ++k){
	
	a1 = 2.0 * (k + 1.0) * (k + a + b + 1.0) * (2.0 * k + a + b);
	a2 = (2.0 * k + a + b + 1.0) * (a * a - b * b) / a1;
	a3 = (2.0 * k + a + b) * (2.0 * k + a + b + 1.0) * (2.0 * k + a + b + 2.0) / a1;
	a4 = 2.0 * (k + a) * (k + b) * (2.0 * k + a + b + 2.0) / a1;
	for (i = 0; i < np; ++i)
	{
	    result_array[i] = (a2 + a3 * x[i]) * pnm1[i] - a4 * pnm2[i];
	    pnm2[i] = pnm1[i];
	    pnm1[i] = result_array[i];
	}
    }


    if (mem_allocated)
    {
	free(pnm1);
    }
    
    return GSL_SUCCESS;
    

}








/** Calculates the derivative of the nth order Jacobi polynomial at an array of np points given by the vector x.
    This function needs two workspaces of size np to do its calculation. If the first one (ws1) is
    NULL, the function will allocate the necessary memory (using malloc) to do the calculation

    \param np Number of points where the polynomial will be calculated
    \param x A vector of length np containing the points where the polynomials should be calculated. \f$-1\le x_i \le 1\f$.
    \param n Order of the Jacobi polynomial whose derivative should be calculated
    \param result_array An array of length np where the result will be stored
    \param a \f$\alpha\f$ parameter of Jacobi polynomial
    \param b \f$\beta\f$ parameter of Jacobi polynomial
    \param ws A pointer to a block of memory 2*np doubles long to be used as workspace. If it is NULL, memory will be allocated using malloc and released at the end
    \return Error code defined in gsl_errno.h or GSL_SUCCESS if everything was fine
 */
int
jac_djacobi_array(int np, const double *x, int n, double * result_array, double a, double b, double *ws)
{
    int code = jac_jacobi_array(np, x, n-1, result_array, a+1.0, b+1.0, ws);
    int i;
    if (code)
	return code;
    
    for (i = 0; i < np; ++i)
	result_array[i] *= 0.5*(a+b+n+1.0);
    return code;
}






/** This function uses an iteration technique to calculate the zeros of a Jacobi polynomial:

  \f[
  P_m^{\alpha,\beta}(x_i) = 0
  \f]

  The algorithm was adapted from the book by Karniadakis and Sherwin. It uses as an initial
  guess the zeros of Chebyshev polynomials.

  \param x Pointer to an array where the zeros of the Jacobi polynomial should be stored
  \param m Order of the polynomial
  \param alpha \f$\alpha\f$ parameter of the Jacobi polynomial
  \param beta \f$\beta\f$ parameter of the Jacobi polynomial

 */
int
jac_jacobi_zeros(double *x, int m, double alpha, double beta)
{
  /* This function returns all of the zeros of a jacobi polynomial. Algorithm taken from B.2 [1];
    
     A good estimate for the roots is the zero for the chebyshev
    polynomial which is readily known
  */
    
    int k, i;
    double r, delta, s, poly;
    const double EPS = 100*GSL_DBL_EPSILON;
    const int MAXITER = 200;
    int iter;
    for (k = 0; k < m; k++)
    {
      /* Initial guess */
	r = -cos( (2*k+1.0)/(2*m) * M_PI);
	if (k > 0) r = (r + x[k-1]) / 2;
	iter = 0;
	do
	{
	    
	    s = 0;
	    
	    for (i=0; i < k; i++)
		s += 1 / (r - x[i]);
	    
	    poly = jac_jacobi(r, m, alpha, beta);
	    delta = - poly / (jac_djacobi(r, m, alpha, beta) - poly*s);
	    r += delta;
	    ++iter;
	    if (iter > MAXITER)
		return GSL_CONTINUE;
	    
	} while (fabs(delta) > EPS);
	
	x[k] = r;
    }
    return GSL_SUCCESS; 
}




src/gsl_utils.c
#include <R.h>
#include <Rmath.h>
/* sys/pow_int.c
 * 
 * Copyright (C) 1996, 1997, 1998, 1999, 2000 Gerard Jungman
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */


double gsl_pow_int(double x, int n)
{
  double value = 1.0;

  if(n < 0) {
    x = 1.0/x;
    n = -n;
  }

  /* repeated squaring method 
   * returns 0.0^0 = 1.0, so continuous in x
   */
  do {
     if(n & 1) value *= x;  /* for n odd */
     n >>= 1;
     x *= x;
  } while (n);

  return value;
}




double gsl_sf_gamma(double x){
  return gammafn(x);
}
double gsl_sf_fact(double x){
  return gsl_sf_gamma(x+1.0);
}





src/gsl_utils.h
/* gsl_utils.h
 * 
 * Copyright (C) 1996, 1997, 1998, 1999, 2000, 2004 Gerard Jungman, Brian Gough
 * Copyright (C) 2006 Paulo Jabardo
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */


/* I don't want the R version to have a dependency on gsl (I use very little of)
   it so I added this small header file to fill in the missing gaps
*/



#ifndef __GSL_UTILS_H_
#define __GSL_UTILS_H_

#include <float.h>
#include <math.h>

/* Code taken from gsl_errno.h */
enum { 
  GSL_SUCCESS  = 0, 
  GSL_FAILURE  = -1,
  GSL_CONTINUE = -2,  /* iteration has not converged */
  GSL_EDOM     = 1,   /* input domain error, e.g sqrt(-1) */
  GSL_ERANGE   = 2,   /* output range error, e.g. exp(1e100) */
  GSL_EFAULT   = 3,   /* invalid pointer */
  GSL_EINVAL   = 4,   /* invalid argument supplied by user */
  GSL_EFAILED  = 5,   /* generic failure */
  GSL_EFACTOR  = 6,   /* factorization failed */
  GSL_ESANITY  = 7,   /* sanity check failed - shouldn't happen */
  GSL_ENOMEM   = 8,   /* malloc failed */
  GSL_EBADFUNC = 9,   /* problem with user-supplied function */
  GSL_ERUNAWAY = 10,  /* iterative process is out of control */
  GSL_EMAXITER = 11,  /* exceeded max number of iterations */
  GSL_EZERODIV = 12,  /* tried to divide by zero */
  GSL_EBADTOL  = 13,  /* user specified an invalid tolerance */
  GSL_ETOL     = 14,  /* failed to reach the specified tolerance */
  GSL_EUNDRFLW = 15,  /* underflow */
  GSL_EOVRFLW  = 16,  /* overflow  */
  GSL_ELOSS    = 17,  /* loss of accuracy */
  GSL_EROUND   = 18,  /* failed because of roundoff error */
  GSL_EBADLEN  = 19,  /* matrix, vector lengths are not conformant */
  GSL_ENOTSQR  = 20,  /* matrix not square */
  GSL_ESING    = 21,  /* apparent singularity detected */
  GSL_EDIVERGE = 22,  /* integral or series is divergent */
  GSL_EUNSUP   = 23,  /* requested feature is not supported by the hardware */
  GSL_EUNIMPL  = 24,  /* requested feature not (yet) implemented */
  GSL_ECACHE   = 25,  /* cache limit exceeded */
  GSL_ETABLE   = 26,  /* table limit exceeded */
  GSL_ENOPROG  = 27,  /* iteration is not making progress towards solution */
  GSL_ENOPROGJ = 28,  /* jacobian evaluations are not improving the solution */
  GSL_ETOLF    = 29,  /* cannot reach the specified tolerance in F */
  GSL_ETOLX    = 30,  /* cannot reach the specified tolerance in X */
  GSL_ETOLG    = 31,  /* cannot reach the specified tolerance in gradient */
  GSL_EOF      = 32   /* end of file */
} ;

/* Code taken from gsl_pow_int.h */


static inline double gsl_pow_2(const double x) { return x*x;   }
static inline double gsl_pow_3(const double x) { return x*x*x; }
static inline double gsl_pow_4(const double x) { double x2 = x*x;   return x2*x2;    }
static inline double gsl_pow_5(const double x) { double x2 = x*x;   return x2*x2*x;  }
inline static double gsl_pow_6(const double x) { double x2 = x*x;   return x2*x2*x2; }
inline static double gsl_pow_7(const double x) { double x3 = x*x*x; return x3*x3*x;  }
inline static double gsl_pow_8(const double x) { double x2 = x*x;   double x4 = x2*x2; return x4*x4; }
inline static double gsl_pow_9(const double x) { double x3 = x*x*x; return x3*x3*x3; }
double gsl_pow_int(double x, int n);



/* Code taken from gsl_sf_result */

struct gsl_sf_result_struct {
  double val;
  double err;
};
typedef struct gsl_sf_result_struct gsl_sf_result;


#define GSL_DBL_EPSILON DBL_EPSILON
#ifndef M_PI
#define M_PI       3.14159265358979323846264338328      /* pi */
#endif


double gsl_sf_gamma(double x);  /* I will redefine it using R's gamma */
double gsl_sf_fact(double x);
#endif /*__GSL_UTILS_H_*/
